### PR TITLE
Separate WAN event diagnostics per publisher

### PR DESCRIPTION
**Pull request by mmedenjak**
Until now, WAN events were shown for all publishers in total, per map
and cache. This is not sufficient for detecting issues when replicating
to multiple clusters (multiple publishers) and some of those publishers
might be having issues while others do not.

EE: https://github.com/hazelcast/hazelcast-enterprise/pull/2153

### CODE REVIEW

**Comment by blazember** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/wan/impl/WanEventCounterPublisherContainer.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+package com.hazelcast.wan.impl;
+
+import com.hazelcast.cache.impl.CacheService;
+import com.hazelcast.map.impl.MapService;
+
+/**
+ * Thread safe container for {@link WanEventCounterContainer}s grouped by
+ * WAN replication name and target group name.
+ */
+class WanEventCounterPublisherContainer {
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/wan/impl/WanEventCounterPublisherContainer.java]: We should come up with a better name. This is a royal PITA to read.

Comment (pveentjer) [reply hazelcast/src/main/java/com/hazelcast/wan/impl/WanEventCounterPublisherContainer.java]: The problem starts with WanEventCounter.

This is not a single counter, it is a set of counters for multiple maps or caches.

Comment (pveentjer) [reply hazelcast/src/main/java/com/hazelcast/wan/impl/WanEventCounterPublisherContainer.java]: Can't we get rid of the WanEventCounterPublisherContainer and let the WanCounter (should be renamed) take care of finding the right counter based on service name and datastructure name?

Comment (pveentjer) [reply hazelcast/src/main/java/com/hazelcast/wan/impl/WanEventCounterPublisherContainer.java]: And we we need to get the counters for every call? Can't we just get those counters once and reuse the reference?

Comment (pveentjer) [reply hazelcast/src/main/java/com/hazelcast/wan/impl/WanEventCounterPublisherContainer.java]: The whole thing with the WanEventCounter is smelly.

**Comment by pveentjer** [comment]
Are EventCounter probes? They really should be.

File: hazelcast/src/main/java/com/hazelcast/wan/impl/WanEventCounterPublisherContainer.java (hunk 1 above)
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/wan/impl/WanEventCounterPublisherContainer.java]: I'll add the caching of the counters. In what way do you consider the `WanEventCounter` smelly? The way I see it, the biggest issue you have with it is that they are part of some "container" objects. This comes from the requirement that a certain event counter should be:
- grouped by publisher. You want to know how many events were sent to one cluster but not the other
- grouped by map/cache. You want to know how many events were sent to one cluster for one map and how many were dropped for a specific cache.
- counters for a specific map/cache need to be removed once a map/cache is destroyed

The final outcome of these requirements is a diagnostics output like this:
```
28-05-2018 13:46:33 WAN[
                          syncState[
                                  status=READY
                                  activeWanConfigName=null
                                  activePublisherName=null
                                  syncedPartitionCount=0]
                          atob[
                                  B[
                                          outboundQueueSize=0
                                          totalPublishedEventCount=1,000
                                          totalPublishLatency=1,293,592
                                          connected=true
                                          paused=false
                                          mapSentEvents-myMap[
                                                  updateCount=1,000
                                                  removeCount=0
                                                  syncCount=0
                                                  droppedCount=0]
                                          mapSentEvents-myMap2[
                                                  updateCount=1,000
                                                  removeCount=0
                                                  syncCount=0
                                                  droppedCount=0]
                                          mapSentEvents-myMap3[
                                                  updateCount=1,000
                                                  removeCount=0
                                                  syncCount=0
                                                  droppedCount=0]]
                                  C[
                                          outboundQueueSize=0
                                          totalPublishedEventCount=1,000
                                          totalPublishLatency=1,293,674
                                          connected=true
                                          paused=false
                                          mapSentEvents-myMap[
                                                  updateCount=1,000
                                                  removeCount=0
                                                  syncCount=0
                                                  droppedCount=0]]]]

28-05-2018 13:46:33 WAN[
                          mapReceivedEvents-myMap[
                                  syncCount=0
                                  updateCount=1,000
                                  removeCount=0]
                          mapReceivedEvents-myMap2[
                                  syncCount=0
                                  updateCount=1,000
                                  removeCount=0]
                          mapReceivedEvents-myMap3[
                                  syncCount=0
                                  updateCount=1,000
                                  removeCount=0]
                          syncState[
                                  status=READY
                                  activeWanConfigName=null
                                  activePublisherName=null
                                  syncedPartitionCount=0]]
```

Can we achieve the same with probes? Do you see some better way of organising the `AtomicLong` counters in the code (not in container classes)?

**Comment by mmedenjak** [comment]
@pveentjer @blazember refactored the WAN counter classes. They are now:
1. `WanEventCounters` -> contains all counters for all publishers, all distributed services and objects
2. `WanPublisherEventCounters` (inner class of `WanEventCounters`) -> all counters for a single publisher
3. `DistributedServiceWanEventCounters` -> all counters for a single publisher, single distributed service
4. `DistributedObjectWanEventCounters` (inner class of `DistributedObjectWanEventCounters`) -> all counters for a single publisher, single distributed object

I can even make the `Distributed*WanEventCounters` classes inner classes of `WanEventCounters` since they are never instantiated outside that class.

I don't think probes are the way to go since we must have this grouping on several levels and we want to render it as such.

Please take a look.

**Comment by mmedenjak** [comment]
Merging this as the PR builder can't get green (missing dependency). @pveentjer I am happy to fix the structure even further if you have any requirements or ideas once we forward-port this to 3.10.x and 3.11. @ahmetmircik @blazember thanks for the reviews!

**Pull request merged**